plugins/inventory/foreman.py

```diff
@@ -238,10 +293,214 @@ def _get_hostvars(self, host, vars_prefix='', omitted_vars=()):
                 hostvars[vars_prefix + k] = v
         return hostvars
 
+    def to_safe(self, word):
+        '''Converts 'bad' characters in a string to underscores
+        so they can be used as Ansible groups
+        >>> ForemanInventory.to_safe("foo-bar baz")
+        'foo_barbaz'
+        '''
+        regex = r"[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", ""))
+
+    def _fetch_params(self):
+        options = ("no", "yes")
+        params = dict()
+        self.want_location = self.get_option('report').get('want_location')
+        self.want_organization = self.get_option('report').get('want_organization')
+        self.want_IPv4 = self.get_option('report').get('want_ipv4')
+        self.want_IPv6 = self.get_option('report').get('want_ipv6')
+        self.want_facts = self.get_option('report').get('want_facts')
+        self.want_host_group = self.get_option('report').get('want_host_group')
+        self.want_hostcollections = self.get_option('report').get('want_hostcollections')
+        self.want_subnet = self.get_option('report').get('want_subnet')
+        self.want_subnet_v6 = self.get_option('report').get('want_subnet_v6')
+        self.want_smart_proxies = self.get_option('report').get('want_smart_proxies')
+        self.want_content_facet_attributes = self.get_option('report').get('want_content_facet_attributes')
+        self.want_host_params = self.get_option('report').get('want_host_params')
+        self.host_filters = self.get_option('host_filters')
+
+        params["Organization".encode('ascii')] = options[self.want_organization]
+        params["Location".encode('ascii')] = options[self.want_location]
+        params["IPv4".encode('ascii')] = options[self.want_IPv4]
+        params["IPv6".encode('ascii')] = options[self.want_IPv6]
+        params["Facts".encode('ascii')] = options[self.want_facts]
+        params["Host Group".encode('ascii')] = options[self.want_host_group]
+        params["Host Collections".encode('ascii')] = options[self.want_hostcollections]
+        params["Subnet".encode('ascii')] = options[self.want_subnet]
+        params["Subnet v6".encode('ascii')] = options[self.want_subnet_v6]
+        params["Smart Proxies".encode('ascii')] = options[self.want_smart_proxies]
+        params["Content Attributes".encode('ascii')] = options[self.want_content_facet_attributes]
+        params["Host Parameters".encode('ascii')] = options[self.want_host_params]
+        if self.host_filters:
+            params["Hosts".encode('ascii')] = self.host_filters
+        return params
+
+    def _use_inventory_report(self):
+        use_inventory_report = self.get_option('foreman').get('use_reports_api')
+        if not use_inventory_report:
+            return False
+        status_url = "%s/api/v2/status" % self.foreman_url
+        result = self._get_json(status_url)
+        foreman_version = (LooseVersion(result.get('version')) >= LooseVersion('1.24.0'))
+        return foreman_version
```

Comment:
technically this returns "is foreman version at least 1.24", not the version itself, so I'd just do
```suggestion
        return (LooseVersion(result.get('version')) >= LooseVersion('1.24.0'))
```

and maybe put the `1.24.0` as a constant like `MINIMUM_FOREMAN_VERSION_FOR_REPORTING_API`?

Reply:
Also, could this check if the `ansible` plugin is installed and return `False` if not?

Probably also good idea to issue warnings if either the version is too old or the plugin not installed but the user requested the feature (then, obviously, my above comment becomes moot)?

Reply:
ah, it *tries* that later:

```python
        try:
            inventory_report_response = self._post_request()
        except Exception:
            self._populate_host_api()
            return
```